README.md
# Time Schedule

React 时间日程安排控件

## DEMO

跟 `antd` 配合使用的DEMO

![Alt Text](https://i.bmp.ovh/imgs/2019/07/def6f025493be745.gif)

## 安装

```bash
npm i time-schedule --save-dev
```

## 使用

```typescript
import 'time-schedule/lib/style/index.css';
import Schedule from 'time-schedule';
```

## 示例

```typescript
<Schedule
  timelineTitle="Time/Name"
  timelineWidth={120}
  startAt="08:00"
  endAt="22:00"
  cellWidth={200}
  maxWidth={850}
  maxHeight={600}
  columns={[
    {key: 'saber', title: 'Saber'},
    {key: 'archer', title: 'Archer'}
  ]}
  dataSource={[]}
/>
```
有关更完整的示例，请参阅`src / example`。


## API

## Schedule Props

参数 | 说明 | 类型 | 默认值
-|-|-|-
className | Scheduler 组件样式类名 | string | -
dataSource | 数据数组 | IDataSourceProps[] | []
startAt | 开始时间，如 '9:00' (HH:mm) | string | '00:00'
endAt | 结束时间，如 '17:00' (HH:mm) | string | '23:50'
timelineTitle | 时间轴显示文字 | string | ReactNode | -
timelineWidth | 时间轴宽度 | number | 80
timelineInterval | 时间区间间隔 (m) 默认值：30（半小时） | number | 30
timelineFormat | 设置时间轴时间格式。配置参考 moment.js | string | 'HH:mm'
columns| 表格列的配置 | IColumnProps[] | []
cellWidth | 单元格宽度 | number | 200
cellHeight | 单元格高度 | number | 30
maxWidth | Scheduler 组件最大宽度 | string \| number | 800
maxHeight | Scheduler 组件最大高度 | string \| number | 500
readonly | 是否只读状态 | boolean | false
autoSize | 是否自动适应最大宽度宽度 | boolean | true

## Event
参数|说明|参数
-|-|-
onSelect | 时间选择回调 | function({rowStartKey: string, rowEndKey: string, col: IColumnProps})
onScroll | 滚动事件 | function(scrollTop: number, scrollLeft: number)
selectedDataFaCC | 数据显示的函数子组件<br />只有 dataSource 变化，才触发子组件更新 | function(data: IDataSourceProps)

## IColumnProps

列描述数据对象

参数 | 说明 | 类型 | 默认值
-|-|-|-
key | React 需要的 key | string | -
title | 列头显示内容	| string \| ReactNode | -
(other fields) | 冗余字段。当调用 onSelect({rowStart, rowEnd, col}) 方法, `col` 将会返回这些冗余字段 | [property: string]: any | -

## IDataSourceProps

数据源描述对象

参数 | 说明 | 类型 | 默认值
-|-|-|-
colKey | IColumnProps 里的 key，用于定位数据的列项 | string | -
rowEndKey | 开始时间，如 '9:00'(HH:mm) | string | -
rowStartKey | 结束时间，如 '17:00'(HH:mm) | string | -
dataKey | 设置显示区域 key 值 | string | -
(other fields) | 冗余字段。当调用 selectedDataFaCC(data) 渲染, `data` 将会返回这些冗余字段 | [property: string]: any | -

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.6](https://github.com/Nelayah/schedule/compare/v1.0.5...v1.0.6) (2019-07-23)


### Bug Fixes

* fix scrollbars style ([17f84d5](https://github.com/Nelayah/schedule/commit/17f84d5))



### [1.0.5](https://github.com/Nelayah/schedule/compare/v1.0.4...v1.0.5) (2019-07-23)


### Bug Fixes

* fix windows browser scrollbars problems ([dfb6e2e](https://github.com/Nelayah/schedule/commit/dfb6e2e))



### [1.0.4](https://github.com/Nelayah/schedule/compare/v1.0.3...v1.0.4) (2019-07-02)



### [1.0.3](https://github.com/Nelayah/schedule/compare/v1.0.2...v1.0.3) (2019-07-02)



### [1.0.2](https://github.com/Nelayah/schedule/compare/v1.0.1...v1.0.2) (2019-07-02)



### 1.0.1 (2019-07-02)


### Bug Fixes

* **calculate:** fix calculate overlapping error bug ([389cef0](https://github.com/Nelayah/schedule/commit/389cef0))


lib/index.js
import * as React from 'react';
import * as R from 'ramda';
import TimeProcessor from './lib/TimeProcessor';
import { Scrollbars } from 'react-custom-scrollbars';
import classnames from 'classnames';
import Columns from './components/columns';
import Provider from './components/provider';
export const prefixCls = 'time-schedule';
export const cellCls = `${prefixCls}-cell`;
export const cellColumnsCls = `${prefixCls}-col`;
export const cellHeaderCls = classnames(cellCls, `${prefixCls}-header-cell`);
export default class default_1 extends React.Component {
    constructor(props) {
        super(props);
        this.header = React.createRef();
        this.row = React.createRef();
        this.cell = React.createRef();
        this.height = 0;
        this.width = 0;
        this.cellWidth = 0;
        this.maxWidth = 0;
        this.maxHeight = 0;
        this.scrollMarker = false;
        // 数据行列滚动
        this.handleBodyScroll = () => {
            const { onScroll } = this.props;
            if (this.scrollMarker)
                return this.scrollMarker = false;
            this.row.current.scrollTop = this.cell.current.view.scrollTop;
            this.header.current.scrollLeft = this.cell.current.view.scrollLeft;
            onScroll && onScroll(this.cell.current.view.scrollTop, this.cell.current.view.scrollLeft);
            this.scrollMarker = true;
        };
        // 固定行列滚动
        this.handleFixedScroll = () => {
            const { onScroll } = this.props;
            if (this.scrollMarker)
                return this.scrollMarker = false;
            this.cell.current.view.scrollTop = this.row.current.scrollTop;
            this.cell.current.view.scrollLeft = this.header.current.scrollLeft;
            onScroll && onScroll(this.row.current.scrollTop, this.header.current.scrollLeft);
            this.scrollMarker = true;
        };
        this.timeProcessor = TimeProcessor.of(props.startAt, props.endAt);
        const serialization = this.timeProcessor.serialize(props.timelineInterval);
        const rows = serialization.map(v => ({
            title: v.format(props.timelineFormat),
            key: v.valueOf()
        }));
        this.width = props.cellWidth * props.columns.length + props.timelineWidth;
        this.height = rows.length * props.cellHeight;
        this.maxHeight = Math.min(props.maxHeight, this.height);
        this.maxWidth = props.autoSize ? props.maxWidth : Math.min(props.maxWidth, this.width);
        this.cellWidth = props.autoSize ? Math.max(Math.floor((props.maxWidth - props.timelineWidth) / props.columns.length), props.cellWidth) : props.cellWidth;
        this.state = {
            isScrollLeft: false,
            isScrollTop: false,
            rows
        };
    }
    componentWillReceiveProps(nextProps) {
        let rows = this.state.rows;
        // @ts-ignore
        const isDiff = R.complement(R.eqProps(R.__, nextProps, this.props));
        // 监听 columns、maxWidth 和 cellWidth 变化
        if (nextProps.columns.length !== this.props.columns.length || isDiff('cellWidth') || isDiff('maxWidth') || isDiff('autoSize') || isDiff('timelineWidth')) {
            this.width = nextProps.cellWidth * nextProps.columns.length + nextProps.timelineWidth;
            this.maxWidth = nextProps.autoSize ? nextProps.maxWidth : Math.min(nextProps.maxWidth, this.width);
            this.cellWidth = nextProps.autoSize ? Math.max(Math.floor((nextProps.maxWidth - nextProps.timelineWidth) / nextProps.columns.length), nextProps.cellWidth) : nextProps.cellWidth;
        }
        // 监听 startAt、endAt、timelineInterval 和 timelineFormat 变化
        if (isDiff('startAt') || isDiff('endAt') || isDiff('timelineInterval') || isDiff('timelineFormat')) {
            this.timeProcessor = TimeProcessor.of(nextProps.startAt, nextProps.endAt);
            const serialization = this.timeProcessor.serialize(nextProps.timelineInterval);
            rows = serialization.map(v => ({
                title: v.format(nextProps.timelineFormat),
                key: v.valueOf()
            }));
        }
        // 监听 rows、cellHeight 和 maxHeight 变化
        if (rows !== this.state.rows || isDiff('cellHeight') || isDiff('maxHeight')) {
            this.height = rows.length * nextProps.cellHeight;
            this.maxHeight = Math.min(nextProps.maxHeight, this.height);
        }
    }
    shouldComponentUpdate(nextProps, nextState) {
        // @ts-ignore
        const isDiffState = R.complement(R.eqProps(R.__, nextState, this.state));
        const isDiffPropsJSON = key => JSON.stringify(nextProps[key]) !== JSON.stringify(this.props[key]);
        const getColumnsKeyString = obj => JSON.stringify(R.pluck('key')(obj.columns));
        const columnsKeyEquals = R.equals(getColumnsKeyString(nextProps), getColumnsKeyString(this.props));
        // @ts-ignore
        const getColumnsTitleSnap = R.compose(R.map(R.when(R.has('toJSON'), v => v.toJSON())), R.pluck('title'), R.prop('columns'));
        const columnsTitleEquals = R.equals(getColumnsTitleSnap(nextProps), getColumnsTitleSnap(this.props));
        const shouldUpdateState = R.keys(this.state).reduce((prev, key) => prev || isDiffState(key), false);
        return isDiffPropsJSON('dataSource') || !columnsKeyEquals || !columnsTitleEquals || shouldUpdateState;
    }
    render() {
        const { columns, timelineTitle, className, dataSource, cellHeight, autoSize, timelineWidth } = this.props;
        const { rows } = this.state;
        const tableWidth = autoSize ? Math.max(this.maxWidth, this.width) : this.width;
        return (React.createElement(Provider, Object.assign({}, this.props, { cellWidth: this.cellWidth }),
            React.createElement("div", { className: classnames(prefixCls, className), style: { maxWidth: this.maxWidth } },
                React.createElement("div", { className: `${prefixCls}-header`, style: { width: tableWidth } },
                    React.createElement("div", { className: classnames(cellHeaderCls, `${prefixCls}-timeline-title`), style: { width: timelineWidth, height: cellHeight, lineHeight: `${cellHeight}px` } }, timelineTitle),
                    React.createElement("div", { className: `${prefixCls}-header-col`, style: { width: this.maxWidth - timelineWidth, height: cellHeight } },
                        React.createElement("div", { ref: this.header, className: `${prefixCls}-header-col-items`, style: { width: this.maxWidth - timelineWidth }, onScroll: this.handleFixedScroll },
                            React.createElement("div", { style: { width: (tableWidth) - timelineWidth } }, columns.map(v => {
                                return (React.createElement("div", { key: v.key, className: cellHeaderCls, style: { width: this.cellWidth, height: cellHeight, lineHeight: `${cellHeight}px` } }, v.title));
                            }))))),
                React.createElement("div", { className: `${prefixCls}-body`, style: { height: this.maxHeight, width: tableWidth } },
                    React.createElement("div", { className: classnames(cellColumnsCls, `${prefixCls}-col-fixed`) },
                        React.createElement("div", { ref: this.row, className: `${prefixCls}-col-fixed-container`, 
                            // 隐藏 多余的滚动条
                            style: { height: this.maxHeight, width: this.maxWidth + 20 }, onScroll: this.handleFixedScroll },
                            React.createElement("div", { className: cellCls, style: { width: timelineWidth, height: cellHeight / 2 } }),
                            rows.map(v => {
                                return (React.createElement("div", { key: v.key, className: cellCls, style: { width: timelineWidth, height: cellHeight, lineHeight: `${cellHeight}px` } }, v.title));
                            }),
                            React.createElement("div", { className: cellCls, style: { width: timelineWidth, height: cellHeight / 2 } }))),
                    React.createElement(Scrollbars, { ref: this.cell, onScroll: this.handleBodyScroll, autoHide: true, style: {
                            marginLeft: timelineWidth,
                            float: 'left',
                            width: this.maxWidth - timelineWidth,
                            height: this.maxHeight
                        } },
                        React.createElement("div", { style: { width: tableWidth - timelineWidth } }, columns.map(v => {
                            return (React.createElement(Columns, { key: v.key, targetCol: v, rows: rows, cellCls: cellCls, dataSource: dataSource.filter(x => x.colKey === v.key) }));
                        })))))));
    }
}
default_1.displayName = 'Time-schedule';
default_1.defaultProps = {
    startAt: '00:00',
    endAt: '23:50',
    timelineInterval: 30,
    timelineWidth: 80,
    timelineFormat: 'HH:mm',
    columns: [],
    dataSource: [],
    cellWidth: 200,
    cellHeight: 30,
    maxWidth: 800,
    maxHeight: 500,
    readonly: false,
    autoSize: true
};
